src/internal/storage/server.go

```diff
@@ -87,3 +104,189 @@ func getOrCreateKey(ctx context.Context, keyStore chunk.KeyStore, name string) (
+	id, err := fileset.ParseID(request.FilesetId)
+	if err != nil {
+		return err
+	}
+	fs, err := s.Filesets.Open(ctx, []fileset.ID{*id})
+	if err != nil {
+		return err
+	}
+	// Compute the intersection of the path range filters.
+	pathRange := &index.PathRange{}
+	for _, f := range request.Filters {
+		switch f := f.Filter.(type) {
+		case *storage.FileFilter_PathRange:
+			// Return immediately if the path ranges are disjoint.
+			if pathRange.Lower != "" && f.PathRange.Upper <= pathRange.Lower ||
+				pathRange.Upper != "" && f.PathRange.Lower >= pathRange.Upper {
+				return nil
+			}
+			if f.PathRange.Lower > pathRange.Lower {
+				pathRange.Lower = f.PathRange.Lower
+			}
+			if pathRange.Upper == "" || f.PathRange.Upper != "" && f.PathRange.Upper < pathRange.Upper {
+				pathRange.Upper = f.PathRange.Upper
+			}
+		}
+	}
+	// Compile regular expressions.
+	var regexes []*regexp.Regexp
+	for _, f := range request.Filters {
+		switch f := f.Filter.(type) {
+		case *storage.FileFilter_PathRegex:
+			regex, err := regexp.Compile(f.PathRegex)
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			regexes = append(regexes, regex)
+		}
+	}
+	return fs.Iterate(ctx, func(f fileset.File) error {
+		path := f.Index().Path
+		for _, r := range regexes {
+			if !r.MatchString(path) {
+				return nil
+			}
+		}
+		w := &writer{
+			server: server,
+			path:   path,
+		}
+		bufW := bufio.NewWriterSize(w, grpcutil.MaxMsgPayloadSize)
+		if err := f.Content(ctx, bufW); err != nil {
+			return err
+		}
+		return errors.EnsureStack(bufW.Flush())
+	}, index.WithRange(pathRange))
+}
+
+type writer struct {
+	server storage.Fileset_ReadFilesetServer
+	path   string
+}
+
+func (w *writer) Write(data []byte) (int, error) {
+	response := &storage.ReadFilesetResponse{
+		Path: w.path,
+		Data: wrapperspb.Bytes(data),
+	}
+	if err := w.server.Send(response); err != nil {
+		return 0, err
+	}
+	return len(data), nil
+}
+
+// TODO: We should be able to use this and potentially others directly in PFS.
+func (s *Server) RenewFileset(ctx context.Context, request *storage.RenewFilesetRequest) (*emptypb.Empty, error) {
+	id, err := fileset.ParseID(request.FilesetId)
+	if err != nil {
+		return nil, err
+	}
+	ttl := time.Duration(request.TtlSeconds) * time.Second
+	if ttl < time.Second {
+		return nil, errors.Errorf("ttl (%d) must be at least one second", ttl)
+	}
+	if ttl > maxTTL {
+		return nil, errors.Errorf("ttl (%d) exceeds max ttl (%d)", ttl, maxTTL)
+	}
+	_, err = s.Filesets.SetTTL(ctx, *id, ttl)
+	if err != nil {
+		return nil, err
+	}
+	return &emptypb.Empty{}, nil
+}
+
+func (s *Server) ComposeFileset(ctx context.Context, request *storage.ComposeFilesetRequest) (*storage.ComposeFilesetResponse, error) {
+	var ids []fileset.ID
+	for _, id := range request.FilesetIds {
+		id, err := fileset.ParseID(id)
+		if err != nil {
+			return nil, err
+		}
+		ids = append(ids, *id)
+	}
+	ttl := time.Duration(request.TtlSeconds) * time.Second
```

Comment:
Is there a default TTL if the request doesn't specify one?

Reply:
There is a default when a fileset is created, but not for renewal. The min and max keep the ttl within a reasonable bound (too low is probably a client bug, and too high keeps the data around longer which is bad for storage and security).